File: internal/game.go
package internal

import (
	"errors"
	"strings"
)

func NewGame(secret string) (*Game, error) {
	if secret == "" {
		return nil, errors.New("secret must be a 5-letter word")
	}
	return &Game{secret: strings.ToUpper(secret)}, nil
}

type Game struct {
	secret string
}
type ClueResult string

const (
	Correct ClueResult = "Correct"
)

type Clue struct {
	Letter string
	Result ClueResult
}

type GuessResult struct {
	Solved bool
	Clues  [5]Clue
}

func (g *Game) Guess(guess string) GuessResult {
	guess = strings.ToUpper(guess)
	if guess == g.secret {
		return GuessResult{
			Solved: true,
		}
	}
	if guess[0] == g.secret[0] {
		return GuessResult{
			Solved: false,
			Clues: [5]Clue{
				{
					Letter: string(g.secret[0]),
					Result: Correct,
				},
			},
		}
	}
	return GuessResult{}
}

File: src/domain/game.ts
type Clue = "CORRECT" | "MISPLACED" | "ABSENT";

export class Game {
  readonly word: string;
  readonly maxGuessesAllowed: number;
  private _isSolved: boolean = false;

  private guessesMade: number = 0;

  public get isSolved(): boolean {
    return this._isSolved;
  }

  constructor(word: string, maxGuessesAllowed: number) {
    this.word = word;
    this.maxGuessesAllowed = maxGuessesAllowed;
  }

  makeGuess(guess: string): Clue[] {
    if (this.guessesMade == this.maxGuessesAllowed) {
      throw new Error("Max guesses exceeded");
    }
    if (guess === this.word) {
      this._isSolved = true;
    }
    this.guessesMade++;

    let clues: Clue[] = [];

    for (let i = 0; i < guess.length; i++) {
      const letter = guess[i];
      if (letter === this.word[i]) {
        clues.push("CORRECT");
        continue;
      }
      if (this.word.includes(letter)) {
        clues.push("MISPLACED");
        continue;
      }
      clues.push("ABSENT");
    }

    return clues;
  }
}
